src/mimeparser.rs

```diff
@@ -217,67 +220,97 @@ impl MimeMessage {
+                    let mut signed_from = Vec::new();
 
-                        // We do not want to allow unencrypted subject in encrypted emails because the user might falsely think that the subject is safe.
-                        // See <https://github.com/deltachat/deltachat-core-rust/issues/1790>.
-                        headers.remove("subject");
+                    // We do not want to allow unencrypted subject in encrypted emails because the user might falsely think that the subject is safe.
+                    // See <https://github.com/deltachat/deltachat-core-rust/issues/1790>.
+                    headers.remove("subject");
 
-                        MimeMessage::merge_headers(
-                            context,
-                            &mut headers,
-                            &mut recipients,
-                            &mut throwaway_from,
-                            &mut list_post,
-                            &mut chat_disposition_notification_to,
-                            &decrypted_mail.headers,
-                        );
+                    // TODO question: Why don't we check if `signatures` is empty here?
+                    // If they are empty (and the mail was incorrectly signed), then do we really
+                    // want to allow "secure-join-fingerprint" and "chat-verified" headers?
+                    // (which we didn't allow in the unencrypted headers above)
+                    MimeMessage::merge_headers(
+                        context,
+                        &mut headers,
+                        &mut recipients,
+                        &mut signed_from,
+                        &mut list_post,
+                        &mut chat_disposition_notification_to,
+                        &decrypted_mail.headers,
+                    );
+                    if let Some(signed_from) = signed_from.first() {
+                        if let Some(from) = from.first() {
+                            if addr_cmp(&signed_from.addr, &from.addr) {
+                                from_is_signed = true;
+                            } else {
+                                // TODO question: there is a From: header in the encrypted &
+                                // signed part, but it doesn't match the outer one.
+                                // This _might_ be because the sender's mail server
+                                // replaced the sending address, e.g. in a mailing list.
+                                // Or it's because someone is doing some replay attack
+                                // - OTOH, I can't come up with an attack where
+                                // someone would send someone else's email from their
+                                // own email address (except for the mentioned AEAP
+                                // attack, which we guard against by checking
+                                // from_is_signed)
```

Comment:
I think I tend to just removing the `else` block and do nothing if the `From:` headers mismatch.